### PR TITLE
Ask again after wrong registration code (bsc#958521)

### PR DESCRIPTION
**Pull request by teclator**
- We will register all valid addons and will continue asking for register codes for the missing ones. You can go back and disable the extension in case you decide to not install it.

### CODE REVIEW

File: package/yast2-registration.changes [hunk 1]
```diff
@@ -1,4 +1,12 @@
 -------------------------------------------------------------------
+Fri Feb 12 12:42:19 UTC 2016 - knut.anderssen@suse.com
+
+- Continue asking for registration codes also after failed
```
Comment (imobach) [inline package/yast2-registration.changes]: I would rephrase it. Maybe something like: `Ask again when an incorrect registration code is provided`.

**Comment by teclator** [comment]
Going back ask again the last EULA code. @lslezak proposed going directly to the addons selection menu, so it's missing yet.

File: src/clients/inst_scc.rb [hunk 2]
```diff
@@ -164,9 +164,11 @@ def get_available_addons
     end
 
     # register all selected addons
+    # back returns directly to the extensions selection
     def register_addons
       return false if init_registration == :cancel
-      registration_ui.register_addons(@selected_addons, @known_reg_codes)
+      ret = registration_ui.register_addons(@selected_addons, @known_reg_codes)
+      (ret == :back) ? :extensions : ret
```
Comment (lslezak) [inline src/clients/inst_scc.rb]: NP: I'd rather use

```ruby
ret = :extensions if ret == :back
ret
```

here as the `()`, `?` and `:` symbols make the line less readable...

**Comment by lslezak** [comment]
LGTM, thanks for the fix! :+1:

File: src/lib/registration/registration_ui.rb [hunk 3]
```diff
@@ -294,13 +294,16 @@ def register_selected_addons(selected_addons, known_reg_codes)
...
-        end
+            register_selected_addon(product, known_reg_codes[product.identifier])
+          end
 
         # remove from selected after successful registration
-        selected_addons.reject! { |selected| selected.identifier == product.identifier }
+        if registered
+          selected_addons.reject! { |selected| selected.identifier == product.identifier }
+        end
+        registered
```
Comment (mvidner) [inline src/lib/registration/registration_ui.rb]: Before, the block returned the result of `reject!`. That seems inconsistent.

File: src/clients/inst_scc.rb [hunk 4]
```diff
@@ -304,12 +306,14 @@ def start_workflow
...
-          next:  "update_autoyast_config"
+          abort:      :abort,
+          extensions: "select_addons",
+          next:       "update_autoyast_config"
         },
         "reregister_addons"      => {
-          abort: :abort,
-          next:  "check"
+          abort:      :abort,
+          extensions: "select_addons",
```
Comment (mvidner) [inline src/clients/inst_scc.rb]: Are you sure about this one? I have actually drawn the dialog transition graph for my feature, and this edge would continue to a different branch. Did you mean "select_addons_rereg" here?

File: src/lib/registration/registration_ui.rb (hunk 3 above)
Comment (lslezak) [inline src/lib/registration/registration_ui.rb]: Previously it returned `true` when the product was successfully removed from the selected list. That should happen in all cases, but it ignored the addon registration result. Now it honors the registration result which looks as a good change to me.

File: src/clients/inst_scc.rb (hunk 4 above)
Comment (teclator) [inline src/clients/inst_scc.rb]: Thanks for that point!! sure i mean select_addons_rereg, i wii fix and test it manually

File: src/clients/inst_scc.rb (hunk 2 above)
Comment (teclator) [inline src/clients/inst_scc.rb]: Ok, i will change it

File: src/lib/registration/registration_ui.rb (hunk 3 above)
Comment (teclator) [inline src/lib/registration/registration_ui.rb]: Exactly, before, it ever deleted the product from the selected addons, so now it register all with good codes and keep not register ones to be filled if you want to.

**Comment by lslezak** [comment]
LGTM
